### PR TITLE
add representation methods and encoders

### PR DESCRIPTION
**Pull request by arfio**
This method adds encoders to be able to easily store the responses in json format.
It also adds repr() method to used models to be able to print and easily see what is contained in a response.

### CODE REVIEW

**Review by MatthewKhouzam — COMMENTED** [review]
Hi Arnaud, I am wondering what the end game is here. Could you describe it in a comment or commit message? It'll allow the reviews to be better directed.

**Comment by arfio** [comment]
The encoders are useful to store objects on disk between multiple calls to a program. This is useful in some cases where you need to recover the previous state of your run.
The representations implementation is very useful to print the results of the requests to know what you got. It is very useful for debugging and testing out things when developping.

**Review by MatthewKhouzam — APPROVED** [review]

**Review by MatthewKhouzam — DISMISSED** [review]

**Review by bhufmann — APPROVED** [review]
LGTM. It has some merge conflicts right now with latest in master branch. @arfio, could you please rebase it?

**Review by MatthewKhouzam — APPROVED** [review]
Merge with might

File: tsp/tsp_client_response.py [hunk 1]
```diff
@@ -47,3 +52,36 @@ def __init__(self, model, status, status_text):
...
+                }
+            elif isinstance(obj.model, OutputDescriptorSet):
+                return {
+                    'model': [OutputDescriptorEncoder().default(output_descriptor) for output_descriptor in obj.model.descriptors],
+                    RESPONSE_STATUS_KEY: obj.status,
+                    STATUS_MESSAGE_KEY: obj.status_text
+                }
+            else:
+                return {
+                    'model': obj.model,
```
Comment (bhufmann) [inline tsp/tsp_client_response.py:83]: Some modules don't have a encoder implemented yet. For example, when running `--list-experiments` and do a json.dumps on it then I get an error:

```
TypeError: Object of type ExperimentSet is not JSON serializable
```

The same for `list-traces`:

```
TypeError: Object of type TraceSet is not JSON serializable
```

**Review by bhufmann — CHANGES_REQUESTED** [review]
Thanks for the contribution. When testing I encountered some issues which I ask you to address.

I had to add some code to test the additions. Is there a way to provide some tests?
(please remember when running pytest, the trace server workspace is wiped out, so run it on a new workspace)

File: tsp/tsp_client_response.py [hunk 2]
```diff
@@ -47,3 +52,36 @@ def __init__(self, model, status, status_text):
...
+
+class TspClientResponseEncoder(json.JSONEncoder):
+    
+    def default(self, obj):
+        if isinstance(obj, TspClientResponse):
+            # Convert TspClientResponse to a dictionary
+            if isinstance(obj.model, GenericResponse):
+                return {
+                    'model': GenericResponseEncoder().default(obj.model),
+                    RESPONSE_STATUS_KEY : obj.status,
```
[the remark this answers is not in the document]
Comment (arfio) [reply tsp/tsp_client_response.py]: Done

File: tsp/tsp_client_response.py [hunk 3]
```diff
@@ -47,3 +52,36 @@ def __init__(self, model, status, status_text):
...
+                    STATUS_MESSAGE_KEY: obj.status_text
+                }
+            elif isinstance(obj.model, OutputDescriptorSet):
+                return {
+                    'model': [OutputDescriptorEncoder().default(output_descriptor) for output_descriptor in obj.model.descriptors],
+                    RESPONSE_STATUS_KEY: obj.status,
```
[the remark this answers is not in the document]
Comment (arfio) [reply tsp/tsp_client_response.py]: Done

File: tsp/tsp_client_response.py [hunk 4]
```diff
@@ -47,3 +52,36 @@ def __init__(self, model, status, status_text):
...
+                    STATUS_MESSAGE_KEY: obj.status_text
+                }
+            else:
+                return {
+                    'model': obj.model,
+                    RESPONSE_STATUS_KEY: obj.status,
```
[the remark this answers is not in the document]
Comment (arfio) [reply tsp/tsp_client_response.py]: Done

File: tsp/tsp_client.py [hunk 5]
```diff
@@ -119,11 +119,33 @@ def open_trace(self, name, path):
...
+        return TspClientResponse(Trace(json.loads(response.content.decode('utf-8'))),
+                                 response.status_code, response.text)
+
+    def open_traces(self, path, name, max_depth, filter):
+        '''
+        Open a trace on the server
+        parameters: Query object
+        :return: :class:`TspClientResponse <Trace>` object
+        :rtype: TspClientResponse
+        '''
```
[the remark this answers is not in the document]
Comment (arfio) [reply tsp/tsp_client.py]: The whole method has been removed as it is not in the TSP specification and max_depth and filter are not used.

File: tsp/xy_model.py [hunk 6]
```diff
@@ -115,35 +106,12 @@ def __init__(self, params):
...
-        else:
-            print(f' Series X-values: {self.x_values}')
-            print(f' Series Y-values: {self.y_values}')
-            print(f' Series tags: {self.tags}')
+    def __repr__(self):
+        return (f'XYSeries(name={self.series_name}, id={self.series_id}'
+           f', x_axis={self.x_axis}' if hasattr(self, 'x_axis') else ''
+           f', y_axis={self.y_axis}' if hasattr(self, 'y_axis') else ''
+           f', x_values={self.x_values}' if hasattr(self, 'x_values') else ''
+           f', y_values={self.y_values}' if hasattr(self, 'y_values') else '')
```
[the remark this answers is not in the document]
Comment (arfio) [reply tsp/xy_model.py:114]: Indeed the if else was nested and the behavior was not as expected. I changed it to be more straightforward and without bugs.

File: tsp/tsp_client_response.py (hunk 1 above)
Comment (arfio) [reply tsp/tsp_client_response.py:95]: I added those encoders and added them to the TspClientResponse.

File: tsp/tsp_client_response.py [hunk 7]
```diff
@@ -47,3 +52,36 @@ def __init__(self, model, status, status_text):
...
+        return 'TspClientResponse(status_code={}, status_text={}, model={})'.format(self.status_code, self.status_text, self.model)
+
+class TspClientResponseEncoder(json.JSONEncoder):
+    
+    def default(self, obj):
+        if isinstance(obj, TspClientResponse):
+            # Convert TspClientResponse to a dictionary
+            if isinstance(obj.model, GenericResponse):
+                return {
+                    'model': GenericResponseEncoder().default(obj.model),
```
[the remark this answers is not in the document]
Comment (arfio) [reply tsp/tsp_client_response.py:71]: I added some conditions to avoid missing attributes.

**Comment by arfio** [comment]
Thank you for the review, there were indeed quite a few bugs to fix. I wonder if there is a good way of validating these functionalities, should we just validate that no exceptions are raised ? or should we validate that the strings generated are equals to test strings ?

This might require a lot of additional tests and code that might change in the future. What do you think ?

**Comment by arfio** [comment]
The CI failed because on an issue with the trace server, it should be fixed when https://github.com/eclipse-tracecompass-incubator/org.eclipse.tracecompass.incubator/pull/89 is merged.

**Comment by bhufmann** [comment]
> Thank you for the review, there were indeed quite a few bugs to fix. I wonder if there is a good way of validating these functionalities, should we just validate that no exceptions are raised ? or should we validate that the strings generated are equal to test strings ?
> 
> This might require a lot of additional tests and code that might change in the future. What do you think ?

It's not easy to verify print-outs. They can change easily. You could verify some content that it's known to be there, but don't compare IDs or order that can change during multiple executions.

Ideally there would be some test cases which could be used to test and helps to know how to use it. It doesn't need to be exhausting tests, but it would help. I had to read the code and google search to know how to try it.

I'm ok to merge it as is, but it would be good to have an "hello world" test case for example purposes.

**Review by bhufmann — APPROVED** [review]
Code looks good to me now. Thanks.

**Comment by bhufmann** [comment]
@MatthewKhouzam are you still ok with this PR after the updates?

**Review by MatthewKhouzam — APPROVED** [review]
lgtm

**Pull request merged**